syntax = "proto3";
option go_package = "pkg/packets";

message Packet{
  oneof msg{
    InputPacket input = 2;
    PositionSyncMessage position_sync = 3;
    PlayerSyncMessage player_sync = 4;
    HealthSyncMessage health_sync = 5;
    AttackSyncMessage attack_sync = 6;
    ManaSyncMessage mana_sync = 7;
    AttackSpeedSyncMessage attack_speed_sync = 8;
    SyncSkillMessage sync_skill = 9;
    ObjectsDestroyedSyncMessage objects_destroyed_sync = 10;
  }
}

message InputPacket{
  oneof input{
    MoveMessage move = 1;
    AttackMessage player_attack = 2;
    ExecuteSkillMessage execute_skill = 3;
  }
}

message MoveMessage{
  float x = 1;
  float y = 2;
}

message AttackMessage{
  uint64 uid = 1;
}

message PlayerUseSkillMessage{
  uint32 skill_pos = 1;
  uint64 target_player = 2;
  float x = 3;
  float y = 4;
}

message PlayerSyncMessage{
  uint64 uid = 1;
  bool self = 2;
}

message ExitSyncMessage{
  uint64 uid = 1;
}

message PositionSyncMessage{
  uint64 id = 1;
  float pos_x = 2;
  float pos_y = 3;
}

message HealthSyncMessage{
  uint64 uid = 1;
  int32 health = 2;
  int32 max_health = 3;
}

message AttackSyncMessage{
  uint64 from = 1;
  uint64 to = 2;
  float cool_down = 3;
}

message ManaSyncMessage{
  uint64 uid = 1;
  int32 mana = 2;
  int32 max_mana = 3;
}

message AttackSpeedSyncMessage{
  uint64 uid = 1;
  float speed = 2;
}

message ExecuteSkillMessage{
  float rotate = 1;
  float pos_x = 2;
  float pos_y = 3;
}

message SyncSkillMessage{
  uint64 uid = 1;
  int32 pos = 2;
  oneof infos{
    SkillInfoMessage skill_info = 3;
  }
}


message SkillInfoMessage{
  uint64 id = 1;
  float angle = 4;
}


message ObjectsDestroyedSyncMessage{
  uint64 id = 1;
}